x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/snapshot/upgrader/SnapshotUpgradeTaskExecutor.java

```diff
@@ -0,0 +1,282 @@
+    private final AnomalyDetectionAuditor auditor;
+    private final JobResultsProvider jobResultsProvider;
+    private volatile ClusterState clusterState;
+    private final Client client;
+
+    public SnapshotUpgradeTaskExecutor(Settings settings,
+                                       ClusterService clusterService,
+                                       AutodetectProcessManager autodetectProcessManager,
+                                       MlMemoryTracker memoryTracker,
+                                       IndexNameExpressionResolver expressionResolver,
+                                       Client client) {
+        super(MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME,
+            MachineLearning.UTILITY_THREAD_POOL_NAME,
+            settings,
+            clusterService,
+            memoryTracker,
+            expressionResolver);
+        this.autodetectProcessManager = autodetectProcessManager;
+        this.auditor = new AnomalyDetectionAuditor(client, clusterService);
+        this.jobResultsProvider = new JobResultsProvider(client, settings, expressionResolver);
+        this.client = client;
+        clusterService.addListener(event -> clusterState = event.state());
+    }
+
+    @Override
+    public PersistentTasksCustomMetadata.Assignment getAssignment(SnapshotUpgradeTaskParams params, ClusterState clusterState) {
+        boolean isMemoryTrackerRecentlyRefreshed = memoryTracker.isRecentlyRefreshed();
+        Optional<PersistentTasksCustomMetadata.Assignment> optionalAssignment = getPotentialAssignment(params, clusterState);
+        if (optionalAssignment.isPresent()) {
+            return optionalAssignment.get();
+        }
+        JobNodeSelector jobNodeSelector = new JobNodeSelector(
+            clusterState,
+            params.getJobId(),
+            MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME,
+            memoryTracker,
+            0,
+            node -> null);
+        return jobNodeSelector.selectNode(
+            Integer.MAX_VALUE,
+            Integer.MAX_VALUE,
+            maxMachineMemoryPercent,
+            isMemoryTrackerRecentlyRefreshed,
+            useAutoMemoryPercentage);
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SnapshotUpgradeTaskParams params, PersistentTaskState state) {
+        SnapshotUpgradeTaskState jobTaskState = (SnapshotUpgradeTaskState) state;
+        SnapshotUpgradeState jobState = jobTaskState == null ? null : jobTaskState.getState();
+        logger.info("[{}] [{}] starting to execute task",
+            params.getJobId(),
+            params.getSnapshotId());
+
+        // This means that we have loaded the snapshot and possibly snapshot was partially updated
+        // This is no good, we should remove the snapshot
+        if (SnapshotUpgradeState.READING_NEW_STATE.equals(jobState)) {
+            deleteSnapshotAndFailTask(task, params.getJobId(), params.getSnapshotId());
+            return;
```

Comment:
If we are assigned to a new node while `reading_new_state`, the snapshot could be corrupted since the files are being overwritten one by one.

Consequently, we audit, log and then delete the snapshot as it is unusable anyways.

It MIGHT be better to add a flag to the snapshot that says "bad snapshot". But, the way to recover here would be to delete the job model state and then restore from an elasticsearch snapshot...Up for debate.